presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java

```diff
@@ -1047,33 +1074,77 @@ private void logShuffleStatsSummary(ShuffleStatsKey key, List<PrestoSparkShuffle
                     DataSize.succinctBytes(bytesPerSecond));
         }
 
-        private static <T> void waitFor(Collection<Future<T>> futures)
-                throws SparkException
+        private long computeNextTimeout()
+                throws TimeoutException
         {
-            try {
-                for (Future<?> future : futures) {
-                    future.get();
+            long timeout = queryCompletionDeadline - System.currentTimeMillis();
+            if (timeout <= 0) {
+                throw new TimeoutException();
+            }
+            return timeout;
+        }
+    }
+
+    private static <T> void waitForActionsCompletionWithTimeout(Collection<JavaFutureAction<T>> actions, long timeout, TimeUnit timeUnit)
+            throws SparkException, TimeoutException
+    {
+        long deadline = System.currentTimeMillis() + timeUnit.toMillis(timeout);
+
+        try {
+            for (JavaFutureAction<?> action : actions) {
+                long nextTimeoutInMillis = deadline - System.currentTimeMillis();
+                if (nextTimeoutInMillis <= 0) {
+                    throw new TimeoutException();
                 }
+                getActionResultWithTimeout(action, nextTimeoutInMillis, MILLISECONDS);
             }
-            catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-                throw new RuntimeException(e);
+        }
+        finally {
+            for (JavaFutureAction<?> action : actions) {
+                if (!action.isDone()) {
+                    action.cancel(true);
+                }
             }
-            catch (ExecutionException e) {
-                propagateIfPossible(e.getCause(), SparkException.class);
-                propagateIfPossible(e.getCause(), RuntimeException.class);
+        }
+    }
 
-                // this should never happen
-                throw new UncheckedExecutionException(e.getCause());
-            }
-            finally {
-                for (Future<?> future : futures) {
-                    if (!future.isDone()) {
-                        future.cancel(true);
+    private static <T> T getActionResultWithTimeout(JavaFutureAction<T> action, long timeout, TimeUnit timeUnit)
+            throws SparkException, TimeoutException
+    {
+        long deadline = System.currentTimeMillis() + timeUnit.toMillis(timeout);
+        try {
+            while (true) {
+                long nextTimeoutInMillis = deadline - System.currentTimeMillis();
+                if (nextTimeoutInMillis <= 0) {
+                    throw new TimeoutException();
+                }
+                try {
+                    return action.get(timeout, timeUnit);
+                }
+                catch (TimeoutException e) {
+                    // guard against spurious wakeup
```

Comment:
Curious, when would this scenario happens?

Reply:
To be honest - I'm just being paranoid, in case the timeout mechanism in Spark is buggy. If you feel strongly about removing it - I can, i don't feel very strong.